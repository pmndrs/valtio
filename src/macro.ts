import type { NodePath } from '@babel/traverse'
import * as t from '@babel/types'
import { createMacro, MacroError } from 'babel-plugin-macros'
// @ts-ignore
import { addNamed } from '@babel/helper-module-imports'

const macro = ({ references }: any) => {
  references.useProxy?.forEach((path: NodePath) => {
    const hook = addNamed(path, 'useSnapshot', 'valtio')
    const proxy = (path.parentPath.get('arguments.0') as any).node
    if (!t.isIdentifier(proxy)) throw new MacroError('no proxy object')
    const snap = t.identifier(`valtio_macro_snap_${proxy.name}`)
    path.parentPath.parentPath.replaceWith(
      t.variableDeclaration('const', [
        t.variableDeclarator(snap, t.callExpression(hook, [proxy])),
      ])
    )
    let inFunction = 0
    path.parentPath.getFunctionParent()?.traverse({
      Identifier(p) {
        if (
          inFunction === 0 && // in render
          p.node !== proxy &&
          p.node.name === proxy.name
        ) {
          p.node.name = snap.name
        }
      },
      Function: {
        enter() {
          ++inFunction
        },
        exit() {
          --inFunction
        },
      },
    })
  })
}

export declare function useProxy<T extends object>(proxyObject: T): void

export default createMacro(macro, { configName: 'valtio' })
