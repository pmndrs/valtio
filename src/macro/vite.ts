// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
import { addNamed } from '@babel/helper-module-imports'
import * as t from '@babel/types'
import {
  createMacroPlugin,
  defineMacro,
  defineMacroProvider,
} from 'aslemammad-vite-plugin-macro'
import { MacroError } from 'babel-plugin-macros'

export const valtioMacro = defineMacro(`useProxy`)
  .withSignature(`<T extends object>(proxyObject: T): void`)
  .withHandler((ctx) => {
    const { path, args } = ctx
    const hook = addNamed(path, 'useSnapshot', 'valtio')
    const proxy = args[0]?.node

    if (!t.isIdentifier(proxy)) throw new MacroError('no proxy object')

    const snap = t.identifier(`valtio_macro_snap_${proxy.name}`)
    path.parentPath?.replaceWith(
      t.variableDeclaration('const', [
        t.variableDeclarator(snap, t.callExpression(hook, [proxy])),
      ])
    )

    let inFunction = 0
    path.parentPath?.getFunctionParent()?.traverse({
      Identifier(p) {
        if (
          inFunction === 0 && // in render
          p.node !== proxy &&
          p.node.name === proxy.name
        ) {
          p.node.name = snap.name
        }
      },
      Function: {
        enter() {
          ++inFunction
        },
        exit() {
          --inFunction
        },
      },
    })
  })

export function provideValtioMacro() {
  return defineMacroProvider({
    id: 'valtio/macro',
    exports: {
      'valtio/macro': {
        macros: [valtioMacro],
      },
    },
  })
}

const macroPlugin = createMacroPlugin({}).use(provideValtioMacro())

export default macroPlugin
