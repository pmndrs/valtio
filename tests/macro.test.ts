import pluginTester from 'babel-plugin-tester'
import plugin from 'babel-plugin-macros'

pluginTester({
  pluginName: 'valtio/macro',
  plugin,
  snapshot: true,
  babelOptions: {
    filename: __filename,
    parserOpts: { plugins: ['jsx'] },
  },
  tests: [
    `
import { useProxy } from '../dist/macro'

const Component = () => {
  useProxy(state)
  return (
    <div>
      {state.count}
      <button onClick={() => ++state.count}>inc</button>
    </div>
  )
}
    `,
  ],
})
