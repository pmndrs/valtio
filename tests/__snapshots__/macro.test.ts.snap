// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`valtio/macro 1. valtio/macro: 1. valtio/macro 1`] = `

import { useProxy } from '../dist/macro'

const Component = () => {
  useProxy(state)
  return (
    <div>
      {state.count}
      <button onClick={() => ++state.count}>inc</button>
    </div>
  )
}

      ↓ ↓ ↓ ↓ ↓ ↓

import { useSnapshot as _useSnapshot } from 'valtio'

const Component = () => {
  const valtio_macro_snap_state = _useSnapshot(state)

  return (
    <div>
      {valtio_macro_snap_state.count}
      <button onClick={() => ++state.count}>inc</button>
    </div>
  )
}


`;
